File: LeetCode/Python/Count Number of Maximum Bitwise-OR Subsets.py
"""
  Count Number of Maximum Bitwise-OR Subsets
  URL: https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets
  Difficulty: Medium
  Topic : Array, Backtracking, Bit Manipulation, Enumeration
"""

import collections
from functools import reduce

class Solution:
    def countMaxOrSubsets(self, nums: List[int]) -> int:
        if len(set(nums)) == 1:
            return (2 ** len(nums)) - 1
        subSet = collections.defaultdict(int)
        def subset(a: list, i: int):
            if i == len(nums):
                return
            for j in range(i,len(nums)):
                temp = a
                temp.append(nums[j])
                key = reduce(lambda r, x: r | x, temp, 0)
                subSet[key] += 1
                subset(temp, j + 1)
                temp.pop()
        subset([],0)
        return sorted(subSet.items(), reverse=True)[0][1]

File: LeetCode/Medium/Count Number of Maximum Bitwise-OR Subsets.swift
/*
  Count Number of Maximum Bitwise-OR Subsets
  URL: https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets
  Difficulty: Medium
  Topic : Array, Backtracking, Bit Manipulation, Enumeration
*/

import Foundation
class Solution {
    func countMaxOrSubsets(_ nums: [Int]) -> Int {
        if Set(nums).count == 1 { return Int(pow(2.0,Double(nums.count))) - 1 }
        
        var dict: [Int:[[Int]]] = [:]
        func subset(a: [Int], i: Int) {
            if i == nums.count { return }
            for j in (i..<nums.count) {
                var temp = a
                temp.append(nums[j])
                let bit = temp.reduce(0) { $0 | $1 }
                if dict[bit] == nil {
                    dict[bit] = [temp]
                } else {
                    dict[bit]?.append(temp)
                }
                subset(a: temp, i: j + 1)
            }
        }
      
        subset(a: [], i: 0)
        return dict.sorted{ $0.key > $1.key }.first!.value.count
    }
}
